src/entropy.rs

```diff
@@ -0,0 +1,339 @@
+    /// is a panic cause for `A`.
+    ///
+    /// ## Remarks
+    ///
+    /// The Kullback-Leibler divergence is a measure used in [Information Theory]
+    /// to describe the relationship between two probability distribution: it only make sense
+    /// when each array sums to 1 with entries between 0 and 1 (extremes included).
+    ///
+    /// By definition, *pᵢ ln(qᵢ/pᵢ)* is set to 0 if *pᵢ* is 0.
+    ///
+    /// [Kullback-Leibler divergence]: https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence
+    /// [Information Theory]: https://en.wikipedia.org/wiki/Information_theory
+    fn kl_divergence(&self, q: &Self) -> Option<A>
+        where
+            A: Float;
+
+    /// Computes the [cross entropy] *H(p,q)* between two arrays,
+    /// where `self`=*p*.
+    ///
+    /// The cross entropy is defined as:
+    ///
+    /// ```text
+    ///            n
+    /// H(p,q) = - ∑ pᵢ ln(qᵢ)
+    ///           i=1
+    /// ```
+    ///
+    /// If the arrays are empty or their lengths are not equal, `None` is returned.
+    ///
+    /// **Panics** if any element in *q* is negative and taking the logarithm of a negative number
+    /// is a panic cause for `A`.
+    ///
+    /// ## Remarks
+    ///
+    /// The cross entropy is a measure used in [Information Theory]
+    /// to describe the relationship between two probability distribution: it only make sense
+    /// when each array sums to 1 with entries between 0 and 1 (extremes included).
+    ///
+    /// The cross entropy is often used as an objective/loss function in
+    /// [optimization problems], including [machine learning].
+    ///
+    /// By definition, *pᵢ ln(qᵢ)* is set to 0 if *pᵢ* is 0.
+    ///
+    /// [cross entropy]: https://en.wikipedia.org/wiki/Cross-entropy
+    /// [Information Theory]: https://en.wikipedia.org/wiki/Information_theory
+    /// [optimization problems]: https://en.wikipedia.org/wiki/Cross-entropy_method
+    /// [machine learning]: https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_error_function_and_logistic_regression
+    fn cross_entropy(&self, q: &Self) -> Option<A>
+    where
+        A: Float;
+}
+
+
+impl<A, S, D> EntropyExt<A, S, D> for ArrayBase<S, D>
+    where
+        S: Data<Elem = A>,
+        D: Dimension,
+{
+    fn entropy(&self) -> Option<A>
+        where
+            A: Float
+    {
+        if self.len() == 0 {
+            None
+        } else {
+            let entropy = self.map(
+                |x| {
+                    if *x == A::zero() {
+                        A::zero()
+                    } else {
+                        *x * x.ln()
+                    }
+                }
+            ).sum();
+            Some(-entropy)
+        }
+    }
+
+    fn kl_divergence(&self, q: &Self) -> Option<A>
+        where
+            A: Float
+    {
+        if (self.len() == 0) | (self.len() != q.len()) {
+            None
+        } else {
+            let kl_divergence: A = self.iter().zip(q.iter()).map(
```

Comment:
I'd prefer to use `ndarray::Zip` because it should have better performance.

Reply:
I have always some problems when using `Zip` - can you help understand what the compiler is complaining about?
![image](https://user-images.githubusercontent.com/20745048/53400205-474b1200-39a5-11e9-9a20-fee5e35dbf60.png)

(`.fold_while` is more appropriate, but I was doing a first step just to get `Zip` there)

Reply:
I got it to work by changing
`let mut temp = ArrayBase::zeros(self.shape());`
into
`let mut temp = Array::zeros(self.raw_dim());`
but I really want to understand the reason behind the extremely different outcome :sweat_smile:

Reply:
Yeah, the issue is that `self.shape()` returns `&[usize]` which is a slice of unspecified length, so when you call `Array::zeros(anything.shape())`, you get a dynamic-dimensional array. (There have been a few issues about this in the `ndarray` repository from confused users, and it would be nice to resolve this raw edge of the API at some point.) `Zip` requires that all arrays have the same dimensionality, so trying to zip together an array with dimension `IxDyn` and arrays with dimension `D` doesn't work (unless you force `D = IxDyn`).

The issue thing is that calling [`ArrayBase::zeros()`](https://docs.rs/ndarray/0.12.1/ndarray/struct.ArrayBase.html#method.zeros) doesn't specify the storage of the array, so the compiler isn't sure what type of owned array to create (e.g. `Array` or `ArcArray`). Writing `Array::zeros()` tells the compiler that you specifically want an `Array` instance.

Reply:
I've created rust-ndarray/ndarray#591 to help improve the docs in this area.